Refactor kubectl label to untie CLI flags from runtime options

jaehnri opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind cleanup

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR unties CLI flags from runtime options in the `kubectl label` command. 

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Ref https://github.com/kubernetes/kubectl/issues/1046

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

jaehnri committed:
Refactor label command to split flags from options

jaehnri committed:
Run AddFlags in NewCmdLabel

dims commented:
If you still need this PR then please rebase, if not, please close the PR

jaehnri commented:
Hi, @dims!
I'll rebase it soon. I need a review here, do you know if I should call someone specifically?

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/label/label.go:
```diff
@@ -140,76 +186,92 @@ func NewCmdLabel(f cmdutil.Factory, ioStreams genericclioptions.IOStreams) *cobr
+	return &LabelOptions{
+		FilenameOptions: f.FilenameOptions,
+
+		PrintFlags: f.PrintFlags,
```
We already have `ToPrinter`, why we are also storing `PrintFlags` in options?

jaehnri replied:
Hmm, your review made me realize the difference between Flags and Options. Did it, thanks!

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/label/label_test.go:
```diff
@@ -696,18 +699,20 @@ pod/foo not labeled
 				options.overwrite = true
 			}
-			if err := options.Complete(tf, cmd, tc.args); err != nil {
+			if err != nil {
```
if options is nil, above assignment will panick `options.ovewrite = true`. Maybe we can change the order and first check `err != nil` before assignment?

jaehnri replied:
Well noticed. Thanks!

ardaguclu reviewed (commented):
Thanks @jaehnri for the PR.

It looks good to me, left a few comments.

ardaguclu commented:
/triage accepted
/priority backlog

jaehnri committed:
Refactor recorder and printer instantiation in label command

- Test overwrite after err validation in label_test
- Remove PrintFlags from LabelOptions
- Remove Recorder from LabelFlags

Signed-off-by: João Henri <joao.henri.cr@gmail.com>

jaehnri commented on staging/src/k8s.io/kubectl/pkg/cmd/label/label.go:
```diff
@@ -50,11 +50,69 @@ const (
+	RecordFlags *genericclioptions.RecordFlags
+	PrintFlags  *genericclioptions.PrintFlags
+
+	Recorder genericclioptions.Recorder
```
Yup, this way we also got rid of the `genericclioptions.NoopRecorder{}` in Flags.
